File: desktop/src/main/retention-default.ts
import fs from 'fs';
import path from 'path';
import os from 'os';

// Seed `cleanupPeriodDays` into ~/.claude/settings.json when the key is
// ABSENT. Claude Code deletes transcript JSONLs whose age exceeds
// cleanupPeriodDays, and its built-in default is 30 days — which silently
// destroys YouCoded's Resume Browser history (2026-06-12 investigation: 221
// named conversations deleted locally). YouCoded is a chat app; users expect
// history to persist, so we seed a year.
//
// Unlike disable-prompt-suggestion.ts (force-overwrites every launch), this
// only writes when the key is missing: an explicit user value — even a
// deliberately short one — is respected.
//
// CC-coupled: `cleanupPeriodDays` is a Claude Code settings contract. See
// youcoded/docs/cc-dependencies.md → "Transcript retention (cleanupPeriodDays)".

export const DEFAULT_CLEANUP_PERIOD_DAYS = 365;

export interface SeedRetentionResult {
  /** True iff settings.json was rewritten (key was absent). */
  changed: boolean;
  /** The value now in effect, or undefined if settings were unreadable. */
  effective: number | undefined;
}

function settingsPath(): string {
  return path.join(os.homedir(), '.claude', 'settings.json');
}

export function seedCleanupPeriodDefault(): SeedRetentionResult {
  const p = settingsPath();
  let settings: Record<string, unknown> = {};
  if (fs.existsSync(p)) {
    try {
      settings = JSON.parse(fs.readFileSync(p, 'utf8'));
    } catch {
      // Do NOT rewrite on parse failure — settings.json carries hooks and
      // enabledPlugins; replacing a corrupt file with just our key would wipe
      // them. (disable-prompt-suggestion.ts writes fresh in this case; that
      // convention is wrong for a low-stakes seeding like this one.)
      return { changed: false, effective: undefined };
    }
  }

  if (typeof settings.cleanupPeriodDays === 'number') {
    return { changed: false, effective: settings.cleanupPeriodDays as number };
  }

  settings.cleanupPeriodDays = DEFAULT_CLEANUP_PERIOD_DAYS;
  fs.mkdirSync(path.dirname(p), { recursive: true });
  // Atomic write (tmp + rename) — same convention as disable-prompt-suggestion.
  const tmp = `${p}.${process.pid}.tmp`;
  fs.writeFileSync(tmp, JSON.stringify(settings, null, 2), 'utf8');
  fs.renameSync(tmp, p);
  return { changed: true, effective: DEFAULT_CLEANUP_PERIOD_DAYS };
}

File: app/src/main/kotlin/com/youcoded/app/runtime/RetentionDefault.kt
package com.youcoded.app.runtime

import org.json.JSONObject
import java.io.File

/**
 * Seed `cleanupPeriodDays` into ~/.claude/settings.json when the key is
 * absent. Claude Code deletes transcripts older than this (its default is 30
 * days when unset), which silently destroys Resume Browser history. Mirrors
 * desktop retention-default.ts — only writes when absent, so an explicit user
 * value is respected. Never rewrites a corrupt settings.json (it carries
 * hooks + enabledPlugins; wiping them is worse than skipping the seed).
 */
class RetentionDefault(homeDir: File) {
    private val settingsFile = File(homeDir, ".claude/settings.json")

    companion object {
        const val DEFAULT_DAYS = 365
    }

    /** @return true iff settings.json was rewritten with the seeded default. */
    fun seedIfAbsent(): Boolean {
        return try {
            // Read existing settings, or start fresh only when the file is
            // genuinely missing. A parse failure below throws into the catch,
            // which deliberately does NOT rewrite — we must not clobber hooks.
            val root = if (settingsFile.exists()) JSONObject(settingsFile.readText()) else JSONObject()
            if (root.has("cleanupPeriodDays")) return false
            root.put("cleanupPeriodDays", DEFAULT_DAYS)
            settingsFile.parentFile?.mkdirs()
            // Atomic write (tmp + rename) — same convention as PromptSuggestionDisabler.
            val tmp = File(settingsFile.parentFile, settingsFile.name + ".tmp")
            tmp.writeText(root.toString(2))
            if (!tmp.renameTo(settingsFile)) settingsFile.writeText(root.toString(2))
            true
        } catch (_: Throwable) {
            false // corrupt/unreadable settings — skip rather than risk wiping hooks
        }
    }
}
